src/ImageSharp.Web/Middleware/ImageSharpMiddleware.cs

```diff
@@ -191,9 +197,18 @@ public class ImageSharpMiddleware
 
         private async Task Invoke(HttpContext httpContext, bool retry)
         {
-            // We expect to get concrete collection type which removes virtual dispatch concerns and enumerator allocations
             CommandCollection commands = this.requestParser.ParseRequestCommands(httpContext);
 
+            // First check for a HMAC token and capture before the command is stripped out.
+            byte[] secret = this.options.HMACSecretKey;
```

Comment:
There's no issue sharing a shared secret across a multitenant install. This is designed to prevent external bad actors not internal resolution of requests. 

You would use a custom `IImageProvider` or authorization middleware that sites in front of this middleware to segregate tenants.